### PR TITLE
Docs: `grep()` using RegExp with global flag `/g`

### LINKED ISSUE #4704: Grep function randomly skips tests
### Description

Probably this issue is my specific, but this behavior is strange  related to grep by regex. I am using mocha programmatically.
when I am using "grep" function.

The the problem is in this method:
`var match = self._grep.test(test.fullTitle());`
(node_modules\mocha\lib\runner.js )

This method outputs false, while it should be true, but only for some 'it' (tests). My describe has 2 it, one is passed, the other not visible because of condition 'false'.

Example:
this one 'it' visible:
'this is some description @testing, @testing-smoke Ensure dashFeeSensitivity exist'
while this 'it' not:
'this is some description @testing, @testing-smoke Ensure dashMaxFloor exist'

bot of them should return 'true', but within mocha, the second example returns 'false':
/@testing |@testing$|@testing-smoke |@testing-smoke$/g.test('this is some description @testing @testing-smoke Ensure dashFeeSensitivity exist')
/@testing |@testing$|@testing-smoke |@testing-smoke$/g.test('this is some description @testing @testing-smoke Ensure dashMaxFloor exist')

for testing purpose (I know my regex),  I changed ` self._grep` on `/@testing |@testing$|@testing-smoke |@testing-smoke$/g` and it works

### Versions

mocha - 8.4.0
node - 14.16.0, 10.15.3
OS - windows 10, 64
@babel/core - 7.8.4

### PR DESCRIPTION
**Pull request by juergba**
### Description

The global flag `/g` in a RegExp expression is behaving weirdly with Mocha's `--grep` option. The selection of tests based on the fullTitle seems wrong as tests get lost, but at the end it's just javascript.
```js
let reIgnore = /a/g;

console.log(reIgnore.test("a")); // true
console.log(reIgnore.test("a")); // false
console.log(reIgnore.test("a")); // true
console.log(reIgnore.test("a")); // false
```

### Description of the Change

Using a flag as `/g` or `/y` doesn't make any sense in this context. 
- improve `Mocha.prototype.grep` for better conversion of `string` into `RegExp`
- clarify documentation

### Alternate Designs

Instead of re-using the same RegExp instance for checking each test, we could construct new instances:
```js
let reIgnore = /a/g;

console.log(new RegExp(reIgnore).test("a")); // true
console.log(new RegExp(reIgnore).test("a")); // true
console.log(new RegExp(reIgnore).test("a")); // true
console.log(new RegExp(reIgnore).test("a")); // true
```
Mocha's `grep` selection is running in any case, with the `--grep` value or with our defaultGrep `/.*/`. Therefore I decided to not apply any fix for this edge case.

### Applicable issues

closes #4704

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -1138,17 +1138,20 @@ Cause Mocha to only run tests matching the given `regexp`, which is internally c
 
 Suppose, for example, you have "api" related tests, as well as "app" related tests, as shown in the following snippet; One could use `--grep api` or `--grep app` to run one or the other. The same goes for any other part of a suite or test-case title, `--grep users` would be valid as well, or even `--grep GET`.
 
+And another option with double quotes: `--grep "groupA|groupB"`.<br>
```
Comment (outsideris) [inline docs/index.md:1141]: `<br>` is not needed here.

**Review by outsideris — APPROVED** [review]

File: docs/index.md (hunk 1 above)
Comment (juergba) [reply docs/index.md:1141]: I tried, there is no `end-of-line` without `<br>` in our markdown file.
